src/Storages/MergeTree/MergeTreeIndexReadResultPool.cpp

```diff
@@ -94,13 +101,346 @@ SkipIndexReadResultPtr MergeTreeSkipIndexReader::read(const RangesInDataPart & p
+        auto * it = roaring::api::roaring64_iterator_create(data.bitmap64);
+        /// There is no way to recover a failed allocation inside roaring64.
+        /// See https://github.com/RoaringBitmap/CRoaring/issues/638
+        if (it == nullptr)
+            throw Exception(ErrorCodes::MEMORY_LIMIT_EXCEEDED, "Failed to allocate roaring64 iterator");
+
+        if (!roaring::api::roaring64_iterator_move_equalorlarger(it, starting_row))
+        {
+            roaring::api::roaring64_iterator_free(it);
+            return false;
+        }
+
+        auto val = roaring64_iterator_value(it);
+        bool has_value = false;
+        while (val < ending_row)
+        {
+            has_value = true;
+            pos[val - starting_row] = 1;
+            if (!roaring::api::roaring64_iterator_advance(it))
+                break;
+            val = roaring64_iterator_value(it);
+        }
+        roaring::api::roaring64_iterator_free(it);
+        return has_value;
+    }
+}
+
+template ProjectionIndexBitmapPtr ProjectionIndexBitmap::createFromRange<UInt32>(UInt32 start, UInt32 end);
+template ProjectionIndexBitmapPtr ProjectionIndexBitmap::createFromRange<UInt64>(UInt64 start, UInt64 end);
+
+template bool ProjectionIndexBitmap::contains<UInt32>(UInt32 value);
+template bool ProjectionIndexBitmap::contains<UInt64>(UInt64 value);
+
+template void ProjectionIndexBitmap::add<UInt32>(UInt32 value);
+template void ProjectionIndexBitmap::add<UInt64>(UInt64 value);
+
+SingleProjectionIndexReader::SingleProjectionIndexReader(
+    std::shared_ptr<MergeTreeReadPoolProjectionIndex> pool,
+    PrewhereInfoPtr prewhere_info,
+    const ExpressionActionsSettings & actions_settings,
+    const MergeTreeReaderSettings & reader_settings)
+    : projection_index_read_pool(std::move(pool))
+    , processor(std::make_unique<MergeTreeSelectProcessor>(
+          std::static_pointer_cast<IMergeTreeReadPool>(projection_index_read_pool),
+          std::make_unique<MergeTreeProjectionIndexSelectAlgorithm>(),
+          std::move(prewhere_info),
+          nullptr /*lazily_read_info*/,
+          actions_settings,
+          reader_settings))
+{
+}
+
+ProjectionIndexBitmapPtr SingleProjectionIndexReader::read(const RangesInDataPart & ranges)
+{
+    bool can_use_32bit_part_offset = ranges.parent_ranges.max_part_offset <= std::numeric_limits<UInt32>::max();
+
+    /// Prepare the read processor with the current part and its read ranges.
+    /// This sets up internal state needed to read the projection data.
+    assert_cast<MergeTreeProjectionIndexSelectAlgorithm &>(*processor->algorithm).preparePartToRead(&ranges);
+    auto res = can_use_32bit_part_offset ? ProjectionIndexBitmap::create32() : ProjectionIndexBitmap::create64();
+
+    /// Start reading chunks from the projection index reader.
+    /// Each chunk contains a column of UInt64 offsets that we insert into the bitmap.
+    while (true)
```

Comment:
Here we should read not more than `max_projection_rows_to_use_projection_index` rows, right?

Reply:
Yes, that’s right. But this limit isn’t too critical. The optimization logic has already enforced it earlier, so even if we read slightly more here (in theory it shouldn't), it’s fine. Ensuring correctness and readability matters more than strictly capping the number of rows read IMO.